addon/test/lib/mock-utils.js

```diff
@@ -0,0 +1,103 @@
+const { Loader, Require, unload,
+        override, descriptor } = require('toolkit/loader');
+const { ensure } = require('sdk/system/unload');
+
+let debug = false;
+
+exports.setDebug = (flag) => debug = flag;
+
+exports.callback = function(name) {
+  let callState = [];
+  let implementCb = null;
+
+  const fn = function() {
+    const args = Array.prototype.slice.call(arguments);
+    args.timestamp = Date.now();
+    if (debug) {
+      console.log(name, args);  // eslint-disable-line no-console
+    }
+    callState.push(args);
+    if (implementCb) {
+      return implementCb.apply(fn, args);
+    }
+  };
+
+  fn.reset = () => {
+    callState = [];
+    implementCb = null;
+  };
+  fn.calls = () => callState;
+  fn.implement = cb => implementCb = cb;
+
+  return fn;
+};
+
+exports.callbacks = function(spec) {
+  const out = {};
+  Object.keys(spec).forEach(prefix => {
+    out[prefix] = {};
+    spec[prefix].forEach(name => {
+      out[prefix][name] = exports.callback(prefix + '.' + name);
+    });
+  });
+  return out;
+};
+
+exports.resetCallbacks = function(callbacks) {
+  Object.keys(callbacks).forEach(prefix => {
+    const items = callbacks[prefix];
+    Object.keys(items).forEach(name => items[name].reset());
+  });
+};
+
+// Build a custom module loader that substitutes mock modules for the test
+// subject's require() calls.
+exports.loader = function(fromTest, subjectPath, modules) {
```

Comment:
It took me a minute to figure out what `fromTest` is and why it's necessary. It might help if the name had "module" in it. :shrug: